src/target/riscv/riscv-013.c

```diff
@@ -606,6 +606,12 @@ static int dmi_op_timeout(struct target *target, uint32_t *data_in,
 					false);
 			if (status == DMI_STATUS_BUSY) {
 				increase_dmi_busy_delay(target);
+				if (dmi_op == DMI_OP_READ) {
```

Comment:
Hello, 

per my understanding of the Debug Spec v0.13, DMI "busy" responses in case of DMI reads would not require any special handling and the code that was already appeared completely sufficient.

Per Debug Spec 0.13, page 66:

> dmi.op == 3: An operation was attempted while a DMI request is still in progress. The data scanned into dmi in this access will be ignored. (...)

Let's assume the following sequence of events:

* New DMI read transaction is issued by dmi_scan() call at line 579
* We obtain DMI_SUCCESS response (dmi.op == 0), therefore we know the transaction has been started correctly and is now in progress (line 585)
* We try to obtain the read result by issuing DMI NOP on line 605
* Let's assume the DMI read transaction is not yet finished (still runs in the background), therefore we obtain BUSY response (dmi.op == 3) to our NOP
* Why to break the loop now? My understanding is that we should keep issuing additional DMI NOPs, and the first NOP that has dmi.op==0 (success) will also carry the data to be read.

Therefore, I do not see the need for the new code. The existing code should cover all BUSY cases.

Please let me know if I missed/misunderstood anything. Thank you.

Reply:
busy is a sticky state, and can only be cleared by writing dmireset. No other operations (including NOPs) will be executed as long as the status is busy.

It is this way so that a debugger can issue a stream of commands without checking for the results, and only check for results at the end. Some hardware (e.g. Olimex adapter) can get much better performance that way. It's unfortunate because in this particular case it means the debugger ends up reading from memory twice, which might have side effects.

Reply:
Hmm... there is possibly some fudging in the spec language that means NOP might not make busy sticky.

Reply:
One thing that's wrong with @JanMatCodasip's summary is the second point. The status we get from the scan where we initiated the read is the result of the previous operation (nop in this case). It's not until the next scan that we learn whether the read took place or not. In my test case it did not, because the read was issued too soon after the previous nop. Does that make sense?

Reply:
Thank you for your replies.

> busy is a sticky state, and can only be cleared by writing dmireset. No other operations (including NOPs) will be executed as long as the status is busy.

I understand the sticky nature of the "busy" state, along with the reasons for it, but forgot to mention it in my fifth point in the sequence. To be precise, I should have said:

* ... we should keep issuing additional DMI NOPs, and the first response that has dmi.op==0 (success) will also carry the data - result of the original "read" operation. For responses with dmi.op==3 (busy), we reset the sticky busy state by writing to dtmcs.dmireset bit and then continue issuing NOPs. The reset of the sticky state is already correctly handled by the original code - in `increase_dmi_busy_delay()` called at line 608.

> One thing that's wrong with @JanMatCodasip's summary is the second point. The status we get from the scan where we initiated the read is the result of the previous operation (nop in this case). It's not until the next scan that we learn whether the read took place or not. 

My understanding of the DTM spec is that: 

* 1) If we issue a "read" transaction (dmi.op==1) and receive success in that very same dmi scan (received dmi.op==0), that indeed means that the _previous_ DMI transaction succeeded, as you said. 
* 2) For our discussion it should not matter what the previous DMI transaction was - it needn't have been just NOP but also read or write. The important thing is that this previous transaction succeeded == it completed and the DTM module is ready for a new transaction.
* 3) So receiving dmi.op==0 (success response) in the same DMI scan where we initiated the read also means that the "read" operation **must have been started and is being processed**, because we know the DTM was idle and ready for transactions at the point when we started the read.

In other words:

If we issue a read, followed too quickly by a NOP, I believe the current code in OpenOCD should already handle it correctly. 

It will detect that the read has been started (by receiving dmi.op==0 in the initial DMI transaction which starts the read). It will then keep issuing NOPs until the result of the already started read operation becomes available (and will proprerly reset the sticky busy state via dmireset each time it encouters it).

So if an issue has been encountered in this realm, it still appears as a potential bug of the target, not a missing logic in OpenOCD. 

Or, please let me know if there is still a piece of the spec that I missed or misunderstood. Thanks.